plugin/evm/limitorders/liquidations.go

```diff
@@ -1,62 +1,97 @@
+func (liq Liquidable) GetUnfilledSize() *big.Int {
+	return big.NewInt(0).Sub(liq.Size, liq.FilledSize)
+}
 
-	toLiquidate := []Liquidable{}
+func (db *InMemoryDatabase) GetLiquidableTraders(market Market, oraclePrice *big.Int) (longPositions []Liquidable, shortPositions []Liquidable) {
+	longPositions = []Liquidable{}
+	shortPositions = []Liquidable{}
+	markPrice := db.lastPrice[market]
 
 	overSpreadLimit := isOverSpreadLimit(markPrice, oraclePrice)
 
 	for addr, trader := range db.GetAllTraders() {
 		position := trader.Positions[market]
-		notionalPosition := position.Size * markPrice
-		var unrealisedPnL float64
-		if position.Size > 0 {
-			unrealisedPnL = notionalPosition - position.OpenNotional
+		notionalPosition := big.NewInt(0).Div(big.NewInt(0).Mul(position.Size, markPrice), BASE_PRECISION) // position.size * markPrice / BASE_PRECISION
+		var unrealisedPnL *big.Int
+		if position.Size.Sign() == 1 {
+			unrealisedPnL.Sub(notionalPosition, position.OpenNotional)
 		} else {
-			unrealisedPnL = position.OpenNotional - notionalPosition
+			unrealisedPnL.Sub(position.OpenNotional, notionalPosition)
 		}
 
-		margin := trader.GetNormalisedMargin() + position.UnrealisedFunding
-		marginFraction := (margin + unrealisedPnL) / notionalPosition
+		margin := big.NewInt(0).Sub(trader.GetNormalisedMargin(), position.UnrealisedFunding)
+		marginFraction := big.NewInt(0).Div(big.NewInt(0).Add(margin, unrealisedPnL), notionalPosition) // margin + unrealisedPnL / notionalPosition
 		if overSpreadLimit {
-			var oracleBasedUnrealizedPnl float64
+			var oracleBasedUnrealizedPnl *big.Int
 
-			oracleBasedNotional := oraclePrice * math.Abs(position.Size)
-			if position.Size > 0 {
-				oracleBasedUnrealizedPnl = oracleBasedNotional - position.OpenNotional
-			} else if position.Size < 0 {
-				oracleBasedUnrealizedPnl = position.OpenNotional - oracleBasedNotional
+			oracleBasedNotional := big.NewInt(0).Div(big.NewInt(0).Mul(oraclePrice, position.Size.Abs(position.Size)), BASE_PRECISION)
+			if position.Size.Sign() == 1 {
+				oracleBasedUnrealizedPnl = big.NewInt(0).Sub(oracleBasedNotional, position.OpenNotional)
+			} else if position.Size.Sign() == -1 {
+				oracleBasedUnrealizedPnl = big.NewInt(0).Sub(position.OpenNotional, oracleBasedNotional)
 			}
 
-			marginFraction = (margin + oracleBasedUnrealizedPnl) / oracleBasedNotional
+			oracleBasedmarginFraction := big.NewInt(0).Div(big.NewInt(0).Add(margin, oracleBasedUnrealizedPnl), oracleBasedNotional)
+			if oracleBasedmarginFraction.Cmp(marginFraction) == 1 {
+				marginFraction = oracleBasedmarginFraction
+			}
 		}
 
-		if marginFraction < maintenanceMargin {
-			toLiquidate = append(toLiquidate, Liquidable{
-				Address: addr,
-				Size: position.Size,
-			})
+		if marginFraction.Cmp(maintenanceMargin) == -1 {
+			liquidable := Liquidable{
+				Address:        addr,
+				Size:           position.Size,
```

Comment:
Only up to 25% of the position can be liquidated at once - but before you make a change here let's discuss on slack because there's a caveat here.